effects mistake with refining inconsistent errors to always-throws

```julia
(NeuralNetworkReachability) pkg> test
...
     Testing Running tests...
Test Summary:                      | Pass  Total  Time
Optional dependencies (not loaded) |    2      2  0.2s

[43958] signal 4 (1): Illegal instruction: 4
in expression starting at /Users/kristoffercarlsson/PkgEvalAnalysis/dev/NeuralNetworkReachability.jl/test/ForwardAlgorithms/forward.jl:169
Verisig at /Users/kristoffercarlsson/PkgEvalAnalysis/dev/NeuralNetworkReachability.jl/src/ForwardAlgorithms/Verisig.jl:26
macro expansion at /Users/kristoffercarlsson/PkgEvalAnalysis/dev/NeuralNetworkReachability.jl/test/ForwardAlgorithms/forward.jl:184 [inlined]
macro expansion at /Users/kristoffercarlsson/julia1.11/usr/share/julia/stdlib/v1.11/Test/src/Test.jl:1700 [inlined]
top-level scope at /Users/kristoffercarlsson/PkgEvalAnalysis/dev/NeuralNetworkReachability.jl/test/ForwardAlgorithms/forward.jl:170
_jl_invoke at /Users/kristoffercarlsson/julia1.11/src/gf.c:2945
jl_toplevel_eval_flex at /Users/kristoffercarlsson/julia1.11/src/toplevel.c:934
jl_toplevel_eval_flex at /Users/kristoffercarlsson/julia1.11/src/toplevel.c:886
ijl_toplevel_eval at /Users/kristoffercarlsson/julia1.11/src/toplevel.c:952 [inlined]
ijl_toplevel_eval_in at /Users/kristoffercarlsson/julia1.11/src/toplevel.c:994
eval at ./boot.jl:428 [inlined]
include_string at ./loading.jl:2486
_jl_invoke at /Users/kristoffercarlsson/julia1.11/src/gf.c:2926
_include at ./loading.jl:2546
include at ./sysimg.jl:38
unknown function (ip: 0x119403739)
_jl_invoke at /Users/kristoffercarlsson/julia1.11/src/gf.c:2926
jl_apply at /Users/kristoffercarlsson/julia1.11/src/./julia.h:2165 [inlined]
do_call at /Users/kristoffercarlsson/julia1.11/src/interpreter.c:126
eval_stmt_value at /Users/kristoffercarlsson/julia1.11/src/interpreter.c:174
eval_body at /Users/kristoffercarlsson/julia1.11/src/interpreter.c:657
eval_body at /Users/kristoffercarlsson/julia1.11/src/interpreter.c:539
```

That is a pretty sick effects mistake you found there:
```
julia> @noinline f() = @assert isdefined(Main, :y)
f (generic function with 4 methods)

julia> g()
ERROR: AssertionError: isdefined(Main, :y)
Stacktrace:
 [1] f()
   @ Main ./REPL[17]:1
 [2] g()
   @ Main ./REPL[15]:1
 [3] top-level scope
   @ REPL[18]:1

julia> y = 2
2

julia> g()
Unreachable reached at 0x7f3e339a4c2b

[2387242] signal 4 (2): Illegal instruction
in expression starting at REPL[20]:1
g at ./REPL[15]:1
```
The `+c` effect is being used to refine this "returns nothing" to "always-throws" via observing that concrete eval may throw, even though that is wrong (consistency does not apply to error paths).
```
 • %1 = < concrete eval > f()::Union{} (+c,+e,!n,+t,+s,!m,+u)
```

--- a/base/compiler/optimize.jl
+++ b/base/compiler/optimize.jl
@@ -830,31 +830,38 @@ function ((; sv)::ScanStmt)(inst::Instruction, lstmt::Int, bb::Int)
 
     stmt_inconsistent = scan_inconsistency!(inst, sv)
 
-    if stmt_inconsistent && inst.idx == lstmt
-        if isa(stmt, ReturnNode) && isdefined(stmt, :val)
+    if stmt_inconsistent
+        if !has_flag(inst[:flag], IR_FLAG_NOTHROW)
+            # Taint :consistent if this statement may raise since :consistent requires
+            # consistent termination. TODO: Separate :consistent_return and :consistent_termination from :consistent.
             sv.all_retpaths_consistent = false
-        elseif isa(stmt, GotoIfNot)
-            # Conditional Branch with inconsistent condition.
-            # If we do not know this function terminates, taint consistency, now,
-            # :consistent requires consistent termination. TODO: Just look at the
-            # inconsistent region.
-            if !sv.result.ipo_effects.terminates
+        end
+        if inst.idx == lstmt
+            if isa(stmt, ReturnNode) && isdefined(stmt, :val)
                 sv.all_retpaths_consistent = false
-            elseif visit_conditional_successors(sv.lazypostdomtree, sv.ir, bb) do succ::Int
-                       return any_stmt_may_throw(sv.ir, succ)
-                   end
-                # check if this `GotoIfNot` leads to conditional throws, which taints consistency
-                sv.all_retpaths_consistent = false
-            else
-                (; cfg, domtree) = get!(sv.lazyagdomtree)
-                for succ in iterated_dominance_frontier(cfg, BlockLiveness(sv.ir.cfg.blocks[bb].succs, nothing), domtree)
-                    if succ == length(cfg.blocks)
-                        # Phi node in the virtual exit -> We have a conditional
-                        # return. TODO: Check if all the retvals are egal.
-                        sv.all_retpaths_consistent = false
-                    else
-                        visit_bb_phis!(sv.ir, succ) do phiidx::Int
-                            push!(sv.inconsistent, phiidx)
+            elseif isa(stmt, GotoIfNot)
+                # Conditional Branch with inconsistent condition.
+                # If we do not know this function terminates, taint consistency, now,
+                # :consistent requires consistent termination. TODO: Just look at the
+                # inconsistent region.
+                if !sv.result.ipo_effects.terminates
+                    sv.all_retpaths_consistent = false
+                elseif visit_conditional_successors(sv.lazypostdomtree, sv.ir, bb) do succ::Int
+                        return any_stmt_may_throw(sv.ir, succ)
+                    end
+                    # check if this `GotoIfNot` leads to conditional throws, which taints consistency
+                    sv.all_retpaths_consistent = false
+                else
+                    (; cfg, domtree) = get!(sv.lazyagdomtree)
+                    for succ in iterated_dominance_frontier(cfg, BlockLiveness(sv.ir.cfg.blocks[bb].succs, nothing), domtree)
+                        if succ == length(cfg.blocks)
+                            # Phi node in the virtual exit -> We have a conditional
+                            # return. TODO: Check if all the retvals are egal.
+                            sv.all_retpaths_consistent = false
+                        else
+                            visit_bb_phis!(sv.ir, succ) do phiidx::Int
+                                push!(sv.inconsistent, phiidx)
+                            end
                         end
                     end
                 end

--- a/test/compiler/effects.jl
+++ b/test/compiler/effects.jl
@@ -1102,7 +1102,7 @@ function f3_optrefine(x)
     @fastmath sqrt(x)
     return x
 end
-@test Core.Compiler.is_consistent(Base.infer_effects(f3_optrefine))
+@test Core.Compiler.is_consistent(Base.infer_effects(f3_optrefine, (Float64,)))
 
 # Check that :consistent is properly modeled for throwing statements
 const GLOBAL_MUTABLE_SWITCH = Ref{Bool}(false)
@@ -1390,10 +1390,13 @@ let; Base.Experimental.@force_compile; func52843(); end
 
 @noinline f53613() = @assert isdefined(@__MODULE__, :v53613)
 g53613() = f53613()
+h53613() = g53613()
 @test !Core.Compiler.is_consistent(Base.infer_effects(f53613))
-@test_broken !Core.Compiler.is_consistent(Base.infer_effects(g53613))
+@test !Core.Compiler.is_consistent(Base.infer_effects(g53613))
 @test_throws AssertionError f53613()
 @test_throws AssertionError g53613()
+@test_throws AssertionError h53613()
 global v53613 = nothing
 @test f53613() === nothing
 @test g53613() === nothing
+@test h53613() === nothing

